pkg/virt-operator/install-strategy/create.go

```diff
@@ -0,0 +1,996 @@
+		}
+		privSCCCopy := privSCC.DeepCopy()
+
+		modified := false
+		users := privSCCCopy.Users
+
+		// remove users from previous
+		if curSccPriv != nil {
+			for _, acc := range curSccPriv.ServiceAccounts {
+				shouldRemove := true
+				// only remove if the target doesn't contain the same
+				// rule, otherwise leave as is.
+				for _, targetAcc := range sccPriv.ServiceAccounts {
+					if acc == targetAcc {
+						shouldRemove = false
+						break
+					}
+				}
+				if shouldRemove {
+					removed := false
+					users, removed = remove(users, acc)
+					modified = modified || removed
+				}
+			}
+		}
+
+		// add any users from target that don't already exist
+		for _, acc := range sccPriv.ServiceAccounts {
+			if !contains(users, acc) {
+				users = append(users, acc)
+				modified = true
+			}
+		}
+
+		if modified {
+			userBytes, err := json.Marshal(users)
+			if err != nil {
+				return false, err
+			}
+
+			data := []byte(fmt.Sprintf(`{"users": %s}`, userBytes))
+			_, err = scc.SecurityContextConstraints().Patch(sccPriv.TargetSCC, types.StrategicMergePatchType, data)
+			if err != nil {
+				return false, fmt.Errorf("unable to patch scc: %v", err)
+			}
+		}
+	}
+
+	if takeUpdatePath {
+		// UPDATE PATH IS
+		// 1. daemonsets - ensures all compute nodes are updated to handle new features
+		// 2. wait for daemonsets to roll over
+		// 3. controllers - ensures controll plane is ready for new features
+		// 4. wait for controllers to roll over
+		// 5. apiserver - toggles on new features.
+
+		// create/update Daemonsets
+		for _, daemonSet := range targetStrategy.daemonSets {
+			err := syncDaemonSet(kv, daemonSet, stores, clientset, expectations)
+			if err != nil {
+				return false, err
+			}
+		}
+
+		// create/update Controller Deployments
+		for _, deployment := range controllerDeployments(targetStrategy) {
+			err := syncDeployment(kv, deployment, stores, clientset, expectations)
+			if err != nil {
+				return false, err
+			}
+
+		}
+
+		// wait for daemonsets
+		for _, daemonSet := range targetStrategy.daemonSets {
+			if !util.DaemonsetIsReady(kv, daemonSet, stores) {
+				log.Log.V(2).Infof("Waiting on daemonset %v to roll over to latest version", daemonSet.GetName())
+				// not rolled out yet
+				return false, nil
+			}
+		}
+		// wait for controller deployments
+		for _, deployment := range controllerDeployments(targetStrategy) {
+			if !util.DeploymentIsReady(kv, deployment, stores) {
+				log.Log.V(2).Infof("Waiting on deployment %v to roll over to latest version", deployment.GetName())
+				// not rolled out yet
+				return false, nil
+			}
+		}
+
+		// create/update API Deployments
+		for _, deployment := range apiDeployments(targetStrategy) {
+			deployment := deployment.DeepCopy()
+			err := syncDeployment(kv, deployment, stores, clientset, expectations)
+			if err != nil {
+				return false, err
+			}
+		}
+	} else {
+		// CREATE/ROLLBACK PATH IS
```

Comment:
well, the problem is that on create, someone can technically post VMs/VMIs as soon as the CRD is registered. So if we perform the apiserver last, then there's a wider possibly someone could introduce an invalid object into the system and that the system would actually react to it.

The thought process here was that if we introduce the api server first, we at least get validation as soon as possible even if the VM/VMI can't be scheduled yet because other control plane components haven't initialized. 

What we really need to do in the future is pull out the webhook registration from the api-server and move it into the operator. That will let us prevent any VM/VMIs from being posted until the api-server is available and "ready". Essentially by posting the admission webhook, we'd be saying no objects can get posted until the admission webhook's backend is responsive.

Reply:
ok, makes sense :+1:

Reply:
> What we really need to do in the future is pull out the webhook registration from the api-server and move it into the operator. That will let us prevent any VM/VMIs from being posted until the api-server is available and "ready". Essentially by posting the admission webhook, we'd be saying no objects can get posted until the admission webhook's backend is responsive.

In #2033 you say that stage 3 is "Automate Zero Downtime Transitions". Wouldn't "prevent any VM/VMIs from being posted until the api-server is available and ready" be downtime?